Add [launchpad].net support

boltgolt opened the pull request:
This adds both a "version" and a "downloads" badge for packages on a launchpad.net PPA using their API. Inspired by #560.

Boltgolt commented:
Fixed!

platan commented on lib/all-badge-examples.js:
```diff
@@ -726,7 +726,15 @@ const allBadgeExamples = [
-      }
+      },
+      {
+        title: 'Launchpad.net PPA downloads',
```
In my opinion world `downloads` is redundant here since whole section is named `Downloads`  (compare with http://shields.io/).

platan commented on server.js:
```diff
@@ -7967,6 +7967,86 @@ cache(function (data, match, sendBadge, request) {
+  var badgeData;
+
+  if (type == 'version') {
+    badgeData = getBadgeData('version', 'invalid');
```
Other version badges uses service name as a badge's left text instead of world `version`.

platan commented on server.js:
```diff
@@ -7967,6 +7967,86 @@ cache(function (data, match, sendBadge, request) {
+    badgeData = getBadgeData('version', 'invalid');
+  }
+  else if (type == 'downloads') {
+    badgeData = getBadgeData('downloads 24h', 'invalid');
```
Other downloads badges uses world `downloads` as a badge's left text. `24h` can be moved to badge's right text as `/day`.

platan commented on services/launchpad/launchpad.tester.js:
```diff
@@ -0,0 +1,35 @@
+  .get('/ppa/version/otto-kesselgulasch/gimp/gimp.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'version',
+    value: Joi.string()
```
I suggest to use one of our existing version validators (https://github.com/badges/shields/blob/master/services/test-validators.js) or creating a new one. `Joi.string()` is too general in my opinion.

platan commented on services/launchpad/launchpad.tester.js:
```diff
@@ -0,0 +1,35 @@
+  .get('/ppa/downloads/otto-kesselgulasch/gimp/gimp.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'downloads 24h',
+    value: Joi.number()
```
What do you think about using `isMetric` validator here? https://github.com/badges/shields/blob/2ded83f8a68cd36102167883d0f9008a7404cb24/services/test-validators.js#L53?

platan commented on server.js:
```diff
@@ -7967,6 +7967,86 @@ cache(function (data, match, sendBadge, request) {
+          }
+
+          // Put the total on the badge
+          badgeData.text[1] = totalCount;
```
`metric` and `downloadCountColor` can be used to format and color downloads value:
https://github.com/badges/shields/blob/2ded83f8a68cd36102167883d0f9008a7404cb24/server.js#L7161-L7162

platan commented on server.js:
```diff
@@ -7967,6 +7967,86 @@ cache(function (data, match, sendBadge, request) {
+
+  // Make a request for packages in the PPA
+  var apiUrl = 'https://api.launchpad.net/1.0/~' + user + '/+archive/' + ppa + '?ws.op=getPublishedBinaries&ws.size=150';
+  request(apiUrl, function(err, res, buffer) {
```
https://github.com/badges/shields/blob/2ded83f8a68cd36102167883d0f9008a7404cb24/lib/error-helper.js#L8
is a great way to check response before parsing response body.

platan commented on server.js:
```diff
@@ -7967,6 +7967,86 @@ cache(function (data, match, sendBadge, request) {
 }));
 
+// Launchpad integration
+camp.route(/^\/launchpad\/ppa\/([^/]+)\/([^/]+)\/([^/]+)\/([^/]+)\.(svg|png|gif|jpg|json)$/,
```
I suggest to replace first group `([^/]+)` with values which are currently expected there `(version|downloads)`.

platan commented on server.js:
```diff
@@ -7967,6 +7967,86 @@ cache(function (data, match, sendBadge, request) {
+  }
+
+  // Make a request for packages in the PPA
+  var apiUrl = 'https://api.launchpad.net/1.0/~' + user + '/+archive/' + ppa + '?ws.op=getPublishedBinaries&ws.size=150';
```
I'm not familiar with Launchpad PPA/archives but I think we should take a look at API https://api.launchpad.net/1.0.html and check if there are simpler ways to get required data. Here we have to load 150 results.

Boltgolt commented:
Implemented all suggestions. Launchpad is not picky with package versioning and only needs a major and minor version number.

Removed the downloads count for now, i need to look at the API a lot more for a workable solution.

Boltgolt commented:
Looks like the owly way to fetch the download count on a binary-by-binary basis, and a single release can have >10 builds.

RedSparr0w commented on server.js:
```diff
@@ -7967,6 +7967,58 @@ cache(function (data, match, sendBadge, request) {
+
+
+      badgeData.text[1] = foundEntry.source_package_version;
+      badgeData.colorscheme = 'blue';
```
Could you could use [`versionText()`](https://github.com/badges/shields/blob/master/server.js#L706) and [`versionColor()`](https://github.com/badges/shields/blob/master/server.js#L707) here.

https://github.com/badges/shields/blob/master/server.js#L706-L707

RedSparr0w commented on services/launchpad/launchpad.tester.js:
```diff
@@ -0,0 +1,21 @@
+    name: 'launchpad',
+    value: Joi.string().regex(/^\d\.\d/)
+  }));
+
```
Could we also add another test for when the `name` doesn't exist.

RedSparr0w commented on services/launchpad/launchpad.tester.js:
```diff
@@ -0,0 +1,21 @@
+  .get('/ppa/version/otto-kesselgulasch/gimp/gimp.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'launchpad',
+    value: Joi.string().regex(/^\d\.\d/)
```
Could you update this to use `\d+` incase the version is something like `12.3.4`

Boltgolt commented:
Hopefully fixed everything suggested. 

In my opinion `2.8.22-ubu18.04.1~ppa` is the right version number to report. The developer made the choice to number it that way and i don't think we should interfere with that.

platan commented on server.js:
```diff
@@ -7967,6 +7967,70 @@ cache(function (data, match, sendBadge, request) {
+
+    try {
+      var data = JSON.parse(buffer);
+      var recentEntry = false;
```
Lines 7994-8023 can be rewritten to:
```js
      const recentEntry = data.entries.filter(entry => entry.source_package_name == name).sort((a, b) => b.date_published.localeCompare(a.date_published))[0];
      if (recentEntry) {
        badgeData.text[1] = versionText(recentEntry.source_package_version);
        badgeData.colorscheme = versionColor(recentEntry.source_package_version);
      } else {
        badgeData.text[1] = 'no package';
        badgeData.colorscheme = 'red';
      }
      sendBadge(format, badgeData);
```
For loop is replaced by filter + sort. 
And 
```js
badgeData.colorscheme =versionText(recentEntry.source_package_version);
```
is replaced by:
```js
badgeData.colorscheme = versionColor(recentEntry.source_package_version);
```

platan commented on server.js:
```diff
@@ -7967,6 +7967,70 @@ cache(function (data, match, sendBadge, request) {
+      sendBadge(format, badgeData);
+
+    } catch (e) {
+        console.log(e);
```
I suggest to remove this `console.log`.

platan commented on services/launchpad/launchpad.tester.js:
```diff
@@ -0,0 +1,36 @@
+  .expectJSONTypes(Joi.object().keys({
+    name: 'launchpad',
+    value: 'invalid'
+  }));
```
Sorting can be tested with such test:
```js
t.create('the newest package')
  .get('/ppa/version/testuser/testppa/testpackage.json')
  .intercept(nock => nock('https://api.launchpad.net')
    .get('/1.0/~testuser/+archive/testppa?ws.op=getPublishedSources&status=Published&source_name=testpackage')
    .reply(200, {
      "total_size": 3,
      "start": 0,
      "entries": [
        {
          "source_package_name": "testpackage",
          "source_package_version": "1.1.0",
          "date_published": "2018-01-25T15:10:16.132432+00:00"
        },
        // below there is the newest version
        {
          "source_package_name": "testpackage",
          "source_package_version": "1.2.0",
          "date_published": "2018-05-28T16:11:26.132567+00:00"
        },
        {
          "source_package_name": "testpackage",
          "source_package_version": "1.0.0",
          "date_published": "2017-05-12T11:40:30.108517+00:00"
        }
      ]
    }
    ))
  .expectJSON({ name: 'launchpad', value: 'v1.2.0' });

```

paulmelnikow commented:
@boltgolt It would be great to get this in! Do you have time to address the comments above?

boltgolt commented:
Not for the next week or 2, i'll see what i can do after that.

paulmelnikow commented:
Great, feel free to reopen when you're ready!
